### PR TITLE
feat: add secrets API

### LINKED ISSUE #184: Add a "secrets" API
### What needs to get done

This API will let applications define words that are "secrets". When printing, these secrets are replaced with a standard password marker (e.g. `******`).

### Why it needs to get done

This is to support build-secrets in craft applications.

### PR DESCRIPTION
**Pull request by tigarmo**
The public API consists of a single method, `Emitter.set_secrets()`, that receives a list of strings and causes said strings to be masked by "*****" in every output generated by the emitter's printer. This includes regular emitter messages, messages generated by log calls, spinner, progress bar and the "streaming brief" prefix.

Fixes #184

- [ ] Have you followed the guidelines for contributing?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox`?

### CODE REVIEW

File: craft_cli/printer.py [hunk 1]
```diff
@@ -416,3 +422,12 @@ def stop(self) -> None:
 
         self.log.close()
         self.stopped = True
+
+    def set_secrets(self, secrets: list[str]) -> None:
+        """Set the list of strings that should be masked out in all outputs."""
+        self.secrets = secrets[:]
```
Comment (sergiusens) [inline craft_cli/printer.py:428]: what's the objective of the `[:]`? do we need a method instead of just setting the attrib?

File: tests/integration/test_messages_integration.py [hunk 2]
```diff
@@ -1437,3 +1437,52 @@ def test_streaming_brief_spinner(capsys, logger, monkeypatch, init_emitter):
         Line("Done stage"),
     ]
     assert_outputs(capsys, emit, expected_err=expected_err, expected_log=expected_log)
+
+
+@pytest.mark.parametrize("output_is_terminal", [True])
+def test_secrets_integrated(capsys, logger, monkeypatch, init_emitter):
```
Comment (sergiusens) [inline tests/integration/test_messages_integration.py:1443]: to document the API expectations, it might be worth to also do a "with pause" and print instead of emit.

File: craft_cli/printer.py (hunk 1 above)
Comment (tigarmo) [reply craft_cli/printer.py:428]: This is to avoid client code modifying the list by accident after the fact, tested [here](https://github.com/canonical/craft-cli/pull/185/files#diff-026c46a62cfc6af61d84f79ea88769a17b59fb54e31346f67fdcc936918afd16R1154)
I suppose `secrets.copy()` would've been clearer, guess old habits die hard

Comment (tigarmo) [reply craft_cli/printer.py]: updated

File: tests/integration/test_messages_integration.py (hunk 2 above)
Comment (tigarmo) [reply tests/integration/test_messages_integration.py:1443]: done; I actually found an issue with the test so thanks for pointing it to me!

**Review by sergiusens — APPROVED** [review]

File: craft_cli/printer.py (hunk 1 above)
Comment (sergiusens) [reply craft_cli/printer.py]: I would have though this would have only been a problem if the method had a default set to a list (where the common practice was to set to None to avoid this)

File: craft_cli/printer.py [hunk 3]
```diff
@@ -416,3 +422,13 @@ def stop(self) -> None:
 
         self.log.close()
         self.stopped = True
+
+    def set_secrets(self, secrets: list[str]) -> None:
+        """Set the list of strings that should be masked out in all outputs."""
+        # Keep a copy, to project against clients modifying the list on accident.
```
Comment (cmatsuoka) [inline craft_cli/printer.py:428]: Is this intended to be "protect"?

File: tests/unit/test_printer.py [hunk 4]
```diff
@@ -1123,3 +1123,100 @@ def test_spinner_silent_on_complete_messages(spinner, monkeypatch):
     time.sleep(0.05)
 
     assert spinner.printer.spinned == []
+
+
+def test_secrets(capsys, log_filepath):
```
Comment (cmatsuoka) [inline tests/unit/test_printer.py:1128]: It could be interesting to add tests for secrets not matching word boundaries (e.g. "term" and "range" to get "apple banana o***** wa*****elon"), multiple repetitions of the secret in the same string, or non-ASCII unicode characters.

Comment (tigarmo) [reply tests/unit/test_printer.py:1128]: these are all very good suggestions, thanks! Added some more tests

**Review by cmatsuoka — APPROVED** [review]
Thanks!
